plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -231,47 +235,122 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName, val isInterface:
+            if (it.symbol.isVal) {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.WildcardPerm()).fillHole(subjectEmbedding)
+            } else {
+                FieldAccessTypeInvariantEmbedding(it, PermExp.FullPerm()).fillHole(subjectEmbedding)
+            }
+        }
 
-        val fieldsProvenInvariants = fields.values
-            .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
-            .flatMap { it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it)) }
+        /**
+         * Adds access to the shared predicate of the fields in the body of the predicate
+         */
+        fun accessFieldsSharedPredicate() = addAllToBody {
+            it.type.sharedPredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it))
+        }
 
-        val accessSuperTypesPredicates = superTypes
-            .filterIsInstance<ClassTypeEmbedding>()
-            .map { PredicateAccessTypeInvariantEmbedding(it.name, PermExp.WildcardPerm()).fillHole(subjectEmbedding) }
+        /**
+         * Adds access to the unique predicate of the unique fields in the body of the predicate
+         */
+        fun accessFieldsUniquePredicate() = addAllToBody {
+            it.isUnique.let { condition ->
+                if (condition) it.type.uniquePredicateAccessInvariant()?.fillHole(PrimitiveFieldAccess(subjectEmbedding, it)) else null
+            }
+        }
 
-        val body = (accessFields + accessFieldPredicates + accessSuperTypesPredicates + fieldsProvenInvariants).toConjunction()
-        return Predicate(name, listOf(subjectEmbedding.toLocalVarDecl()), body.pureToViper(toBuiltin = true))
+        /**
+         * Adds access to super-types predicates in the body of the predicate
+         */
+        fun accessSuperTypesPredicates(perm: PermExp, name: ClassTypeEmbedding.() -> MangledName) = body.addAll(
+            type.classSuperTypes.map { PredicateAccessTypeInvariantEmbedding(it.name(), perm).fillHole(subjectEmbedding) }
+        )
+
+        /**
+         * Adds the proven invariants in the body of the predicate
+         */
+        fun fieldsProvenInvariants() = addAllToBodyFlat {
+            it.type.provenInvariants().fillHoles(PrimitiveFieldAccess(subjectEmbedding, it))
+        }
+
+        private inline fun addAllToBody(transform: (UserFieldEmbedding) -> ExpEmbedding?) {
+            body.addAll(fields.mapNotNull(transform))
+        }
+
+        private inline fun addAllToBodyFlat(transform: (UserFieldEmbedding) -> List<ExpEmbedding>) {
+            body.addAll(fields.flatMap(transform))
+        }
+    }
+
+    private fun initSharedPredicate() = ClassPredicateBuilder(this).build(name) {
+        filterFields { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+        accessFields()
+        accessFieldsSharedPredicate()
+        accessSuperTypesPredicates(PermExp.WildcardPerm()) { name }
+        fieldsProvenInvariants()
```

Comment:
To be honest, I don't really like this design.  The benefit of a builder is that you should be able to specify what the predicate should contain in an easy-to-read fashion.  Here, you still need to know what all these functions do; the builder doesn't really add anything.

I think one problem we're having is that our naming isn't great: `proven`, `access`, it is all a bit vague.  Maybe it would be good to take a step back, identify the permissions we have, and then write a builder that lets you select fields/supertypes and then add the relevant permissions for them?